### PR TITLE
Ap 2242 gateway evidence upload page

### PR DESCRIPTION
**Pull request by eloisetait3**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/AP-2242)

Gateway Evidence upload page created to upload additional supporting evidence if required. 

## Checklist

Before you ask people to review this PR:

- [x] Tests and rubocop should be passing: `bundle exec rake`
- [x] Github should not be reporting conflicts; you should have recently run `git rebase master`.
- [x] There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- [x] The PR description should say what you changed and why, with a link to the JIRA story.
- [x] You should have looked at the diff against master and ensured that nothing unexpected is included in your changes.
- [x] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: app/controllers/providers/gateway_evidences_controller.rb [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      render :show
+    end
+
+    private
+
+    def save_or_continue
+      if form.files?
+        save_continue_or_draft(form)
+      else
+        legal_aid_application.reload
```
Comment (eloisetait3) [inline app/controllers/providers/gateway_evidences_controller.rb:30]: Is this a good solution to stop the application saving an obsolete gateway evidence record if there are no files attached?

Comment (stephenrichards) [reply app/controllers/providers/gateway_evidences_controller.rb:30]: I don't think you need to reload it here.  From what I can see, the `GatewayEvidence` record is only created when you've pressed the upload button for the first time. So I think, when you're here and `form.files?` returns false, then the `GatewayEvidence` record hasn't been built and attached to the `LegalAidApplication` and you could just return nil.  Or am I missing something?

File: app/forms/providers/gateway_evidence_form.rb [hunk 2]
```diff
@@ -0,0 +1,147 @@
+module Providers
+  class GatewayEvidenceForm # rubocop:disable Metrics/ClassLength
+    include BaseForm
+
```
Comment (stephenrichards) [inline app/forms/providers/gateway_evidence_form.rb:4]: A lot of this seems identical to the `StatementOfCaseForm`, in whcih case I think you could extract the common code out into a `BaseFileUploader` from, then derive the `StatementOfCaseForm` and this form from it.  It seems like more work now, but if ever we have to change it, it means a change in just one place.

File: app/services/flow/flows/provider_merits.rb [hunk 3]
```diff
@@ -124,6 +124,10 @@ class ProviderMerits < FlowSteps # rubocop:disable Metrics/ClassLength
         },
         merits_task_lists: {
           path: ->(application) { urls.providers_legal_aid_application_merits_task_list_path(application) },
+          forward: :gateway_evidences
+        },
```
Comment (stephenrichards) [inline app/services/flow/flows/provider_merits.rb:128]: From reading the ticket, gateway evidence is only required if multiple proceedings have been selected, so I think here we want to go straight on to check merits answers if there is only one proceeding.

**Review by stephenrichards — CHANGES_REQUESTED** [review]
A good job on a complex task, but there are a couple of things I would suggest doing.  Am happy to discuss.

File: app/forms/providers/gateway_evidence_form.rb (hunk 2 above)
Comment (eloisetait3) [reply app/forms/providers/gateway_evidence_form.rb:4]: Yes great minds... I had written that as a ticket to do off the back of this. Thought I'd get this up and working with the right tests + functionality, and then look at refactoring the forms/uploader together later

**Review by stephenrichards — APPROVED** [review]

**Pull request merged**